src/tacacs/bash_tacplus/bash_tacplus.c

```diff
@@ -0,0 +1,483 @@
+        output_debug("TACACS+ debug enabled.\n");
+    }
+}
+
+/*
+ * Load tacacs config.
+ */
+void check_and_load_changed_tacacs_config()
+{
+    struct stat attr;
+    // get config file stat, check if file changed
+    stat(tacacs_config_file, &attr);
+    char date[36];
+    strftime(date, 36, "%d.%m.%Y %H:%M:%S", localtime(&(attr.st_mtime)));
+    if (difftime(attr.st_mtime, config_file_attr.st_mtime) == 0) {
+        output_debug("tacacs config file not change: last modified time: %s.\n", date);
+        return;
+    }
+
+    output_debug("tacacs config file changed: last modified time: %s.\n", date);
+
+    // config file changed, update file stat and reload config.
+    config_file_attr = attr;
+
+    // load config file
+    load_tacacs_config();
+}
+
+/*
+ * Tacacs plugin initialization.
+ */
+void plugin_init()
+{
+    // get config file stat, will use this to check config file changed
+    stat(tacacs_config_file, &config_file_attr);
+
+    // load config file: tacacs_config_file
+    load_tacacs_config();
+
+    output_debug("tacacs plugin initialized.\n");
+}
+
+/*
+ * Tacacs plugin release.
+ */
+void plugin_uninit()
+{
+    output_debug("tacacs plugin un-initialize.\n");
+}
+
+/*
+ * Check if current user is local user.
+ */
+int is_local_user(char *user)
+{
+    if (user == unknown_username) {
+        // for unknown user name, when tacacs enabled, always authorization with tacacs.
+        return IS_REMOTE_USER;
+    }
+
+    struct passwd pwd;
+    struct passwd *pwdresult;
+    char *buf;
+    size_t bufsize = sysconf(_SC_GETPW_R_SIZE_MAX);
+    if (bufsize == -1) {
+        bufsize = DEFAULT_SC_GETPW_R_SIZE_MAX;
+    }
+
+    buf = malloc(bufsize);
+    if (buf == NULL) {
+       output_error("failed to allocate getpwnam_r buffer.\n");
+       return ERROR_CHECK_LOCAL_USER;
+    }
+
+    int s = getpwnam_r(user, &pwd, buf, bufsize, &pwdresult);
+    int result = IS_LOCAL_USER;
+    if (pwdresult == NULL) {
+        if (s == 0)
+            output_error("get user information user failed, user: %s not found\n", user);
+        else {
+            output_error("get user information failed, user: %s, errorno: %d\n", user, s);
+        }
+        
+        result = ERROR_CHECK_LOCAL_USER;
+    }
+    else if (strncmp(pwd.pw_gecos, REMOTE_USER_GECOS_PREFIX, strlen(REMOTE_USER_GECOS_PREFIX)) == 0) {
+        output_debug("user: %s, UID: %d, GECOS: %s is remote user.\n", user, pwd.pw_uid, pwd.pw_gecos);
+        result = IS_REMOTE_USER;
+    }
+    else {
+        output_debug("user: %s, UID: %d, GECOS: %s is local user.\n", user, pwd.pw_uid, pwd.pw_gecos);
+        result = IS_LOCAL_USER;
+    }
+
+    free(buf);
+    return result;
+}
+
+/*
+ * Get user name.
+ */
+char* get_user_name(char *user)
+{
+    if (user != NULL) {
+        return user;
+    }
+
+    // uid is the real user id: https://man7.org/linux/man-pages/man2/geteuid.2.html
+    output_debug("Login user name is empty, try get user name by euid.\n");
```

Comment:
>Login user name is empty, try get user name by euid [](http://example.com/codeflow?start=18&length=51)

Is it better to check uid for all the cases? Is it heavy?

Reply:
The user name is from bash, and in the bash code, the user name is already from uid. so it's not necessary to get the user name again of it's not null.